### PR TITLE
Create Best Practices and FAQ.md closes  #405

### LINKED ISSUE #405: [Guide] Best Practices and/or FAQ
I was thinking about some common bugs (and one's that I've run into) this morning.

It would be nice to have a "Best Practices" / "Tips" / "FAQ" guide in the docs.

-------------------------

Some initial notes:

- Assert `isKindOfClass:` on the object you receive in `-didUpdateToObject:` in your section controllers (Makes it easy to track down simple mistakes in `ListAdapaterDataSource` implementation and non-unique diff identifiers)
- Make sure your diff identifiers are unique
- Look up questions in GH issues and see if there's anything else worth calling out

### PR DESCRIPTION
**Pull request by jessesquires**
## Changes in this pull request

Initial best practices / FAQ. Closes  #405.

### CODE REVIEW

**Comment by jessesquires** [comment]
### Will continue to add more to this PR. This is just the initial guide.

**Comment by jessesquires** [comment]
Ok. I think I've got everything.

File: Guides/Best Practices and FAQ.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+```objective-c
+- (void)didUpdateToObject:(id)object {
+    NSParameterAssert([object isKindOfClass:[MyModelClass class]]);
+    _myModel = object;
+}
+```
+
+- Make sure your [`-diffIdentifier`](https://instagram.github.io/IGListKit/Protocols/IGListDiffable.html#/c:objc(pl)IGListDiffable(im)diffIdentifier) implementation returns a **unique identifier** for each object.
+
+- We highly recommend using single-item sections when possible. That is, each section only has 1 cell. Each section controller manages a single model and a single cell. This gives you the greatest amount of flexibility, modularity, and re-use for your components.
```
Comment (rnystrom) [inline Guides/Best Practices and FAQ.md]: Do we want to recommend this? I think your point is not to use a single section controller w/ tons of variable items, right? We probably don't want to discourage multi-cell section controllers in general.

File: Guides/Best Practices and FAQ.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+**The documentation and examples have `<X>` feature or changes, but I don't have it in my version. Why?**
+
+This feature is on the `master` branch only and hasn't been officially tagged and [released](https://github.com/Instagram/IGListKit/releases). If you need to, you can [install from the `master` branch](https://instagram.github.io/IGListKit/installation.html).
+
+**Does `IGListKit` work with...?**
+
+- Core Data ([#460](https://github.com/Instagram/IGListKit/issues/460), [#461](https://github.com/Instagram/IGListKit/issues/461))
+- AsyncDisplayKit ([AsyncDisplayKit/#2942](https://github.com/facebook/AsyncDisplayKit/pull/2942))
+- ComponentKit
```
Comment (rnystrom) [inline Guides/Best Practices and FAQ.md]: Let's add a link to @ocrickard's work https://github.com/ocrickard/IGListKit-ComponentKit/blob/master/IGListKit-ComponentKit/CKComponentSectionController.mm

File: Guides/Best Practices and FAQ.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- React Native
+- Reactive Cocoa
```
Comment (rnystrom) [inline Guides/Best Practices and FAQ.md]: Should we add RxSwift?

File: Guides/Best Practices and FAQ.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+Yes.
+
+**Does `IGListKit` work with `UITableView`?**
+
+No, but you can install the [diffing subspec via CocoaPods]((https://instagram.github.io/IGListKit/installation.html)).
```
Comment (rnystrom) [inline Guides/Best Practices and FAQ.md]: Is the markdown messed up here? The link doesn't show up in the preview

https://github.com/Instagram/IGListKit/blob/0e1525966448c5c8c25a78a686731f8ae575550f/Guides/Best%20Practices%20and%20FAQ.md

File: Guides/Best Practices and FAQ.md (hunk 1 above)
Comment (jessesquires) [inline Guides/Best Practices and FAQ.md]: 👍 

Right -- I'll remove, unless you have a suggestion on wording here

File: Guides/Best Practices and FAQ.md (hunk 2 above)
Comment (jessesquires) [inline Guides/Best Practices and FAQ.md]: I was looking for that 😄

File: Guides/Best Practices and FAQ.md (hunk 4 above)
Comment (jessesquires) [inline Guides/Best Practices and FAQ.md]: oops

File: Guides/Best Practices and FAQ.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+```objective-c
+- (void)didUpdateToObject:(id)object {
+    NSParameterAssert([object isKindOfClass:[MyModelClass class]]);
+    _myModel = object;
+}
+```
+
+- Make sure your [`-diffIdentifier`](https://instagram.github.io/IGListKit/Protocols/IGListDiffable.html#/c:objc(pl)IGListDiffable(im)diffIdentifier) implementation returns a **unique identifier** for each object.
+
+- We highly recommend using single-item sections when possible. That is, each section controller manages a single model (which may have one or multiple cells). This gives you the greatest amount of flexibility, modularity, and re-use for your components.
```
Comment (jessesquires) [inline Guides/Best Practices and FAQ.md]: @rnystrom - how's this?

**Comment by jessesquires** [comment]
Updated.

**Comment by jessesquires** [comment]
@Sherlouk 

> May it be an idea to add a link to this in the issue template?

+1

> Random question but is isKindOfClass the same as doing (in Swift) if sectionController is DemoSectionController { ...

Yep! We should probably add the Swift equivalent

**Pull request closed without merging**